src/core/lib/iomgr/ev_poll_and_epoll_posix.c

```diff
@@ -547,10 +547,13 @@ static int set_ready_locked(grpc_exec_ctx *exec_ctx, grpc_fd *fd,
 
 static void fd_shutdown(grpc_exec_ctx *exec_ctx, grpc_fd *fd) {
   gpr_mu_lock(&fd->mu);
-  GPR_ASSERT(!fd->shutdown);
```

Comment:
Like we talked about - would prefer this to be in tcp_endpoint directly to simplify possible fd implementations (there will be many).